result.go

```diff
@@ -546,17 +546,46 @@ func (q Query) ToMapByT(result interface{}, keySelectorFn interface{}, valueSele
 	q.ToMapBy(result, keySelectorFunc, valueSelectorFunc)
 }
 
-// ToSlice iterates over a collection and populates the result slice with elements.
-// Collection elements must be assignable to the slice's element type.
-// ToSlice doesn't empty the result slice before populating it.
-func (q Query) ToSlice(result interface{}) {
-	res := reflect.ValueOf(result)
+// ToSlice iterates over a collection and saves the results in the slice pointed
+// by v. It starts saving elements starting from index 0, therefore it will
+// overwrite the slice.
+//
+// If the slice pointed by v has sufficient capacity, v will be pointed to a
+// resliced slice. If it does not, a new underlying array will be allocated and
+// v will point to it.
+func (q Query) ToSlice(v interface{}) {
+	res := reflect.ValueOf(v)
 	slice := reflect.Indirect(res)
 	next := q.Iterate()
-
+	index := 0
 	for item, ok := next(); ok; item, ok = next() {
-		slice = reflect.Append(slice, reflect.ValueOf(item))
+		if index >= slice.Len() {
+			slice = grow(slice, 1)
```

Comment:
@cleitonmarx Sorry it's taking too many iterations but I think I found a critical perf issue:

 I know that we don't know the result side in advance, but it looks like this code is not super optimal.

1. `slice.Len()`  (which contains a type switch == slow) here
2. `grow(slice,1)`

every time we add an element (`N` times). I think this algorithm is inefficient. How about the following one:

- we read the cap of slice `s`, and reslice the `s` to its full capacity (`cap(s)`). so now `len(s)==cap(s)`.
- store this `len:=s.Len()`  into a variable.
- we go into the for loop and to add elements
    - if `index >= len`, we grow the slice by a factor of 2 (meaning, double the slice)
        - if we do this, then **reslice** the `s` to the new `s.Cap()` and update `len = s.Cap()` as well
    - else
        - just set the element at `index`
- before returning, **reslice** s to `0:index`

This way at the end, we'll still keep doubling the cap oft he given slice --however, we're not increasing len one at a time. in this algorithm, it will be always `len(s) == cap(s)` before the last step which resizes. By doing that:
- you actually call `grow()` only a few times (logarithmic)
- no more `N` calls to `s.Len()` as you store `len`/`cap` in variables as you grow
- Perhaps `grow()` can also return the new capacity, this way you don't actually ever need to  call `.Cap()` in the algorithm above.

Can we implement `ToSlice()` like this?

That said, I'd ❤️  to see how this current algorithm vs the one I proposed above behaves in a benchmark.